File: backend/app/schemas/student_learning_loop.py
# Status: real

"""Student-owned path-replan and resource-feedback API contracts."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator


ReplanStatus = Literal["pending", "deferred", "accepted", "reverted", "expired"]
ReplanDecision = Literal["accept", "defer", "revert"]
RecommendationStatus = Literal[
    "scheduled", "accepted", "deferred", "rejected", "superseded", "feedback_received", "completed"
]
FeedbackKind = Literal["too_difficult", "too_shallow", "missing_example", "want_diagram", "want_practice"]
FeedbackStatus = Literal[
    "submitted", "retry_requested", "regenerated", "provider_unavailable", "failed", "rejected"
]


class PathReplanCreateRequest(BaseModel):
    """The caller may cite one owned trigger, or let the service choose the latest owned event."""

    trigger_event_id: UUID | None = None
    assessment_workflow_run_id: UUID | None = None

    @model_validator(mode="after")
    def _at_most_one_trigger(self) -> "PathReplanCreateRequest":
        if self.trigger_event_id is not None and self.assessment_workflow_run_id is not None:
            raise ValueError("一次重规划只能引用一个触发来源")
        return self


class PathReplanDecisionRequest(BaseModel):
    decision: ReplanDecision
    reason: str | None = Field(default=None, max_length=400)


class PathTaskChangeDTO(BaseModel):
    action: Literal["added", "retained"]
    title: str
    knowledge_point: str | None = None
    status: Literal["todo", "active", "done", "blocked"]
    expected_minutes: int = Field(ge=0)


class PathVersionDTO(BaseModel):
    id: UUID
    version_no: int = Field(ge=1)
    kind: Literal["baseline", "replan", "revert"]
    state: Literal["active", "historical"]
    title: str
    summary: str
    diff: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class PathReplanCandidateDTO(BaseModel):
    id: UUID
    status: ReplanStatus
    source_version_no: int = Field(ge=1)
    accepted_version_no: int | None = Field(default=None, ge=1)
    trigger_label: str
    trigger_at: datetime | None = None
    reason_code: str
    reason_text: str
    affected_knowledge_point: str | None = None
    expected_minutes: int = Field(ge=0)
    changed_tasks: list[PathTaskChangeDTO] = Field(default_factory=list)
    source_boundary: str
    created_at: datetime
    updated_at: datetime


class ResourceRecommendationDTO(BaseModel):
    id: UUID
    resource_id: UUID
    title: str
    resource_type: str
    knowledge_point: str | None = None
    status: RecommendationStatus
    scheduled_at: datetime
    rationale: str
    source_boundary: str
    created_at: datetime


class ResourceRecommendationDecisionRequest(BaseModel):
    decision: Literal["accept", "defer", "reject", "complete"]
    reason: str | None = Field(default=None, max_length=400)


class ResourceFeedbackRequest(BaseModel):
    feedback_kinds: list[FeedbackKind] = Field(min_length=1, max_length=5)
    comment: str | None = Field(default=None, max_length=500)
    recommendation_id: UUID | None = None
    provider: str | None = Field(default=None, max_length=64)
    model: str | None = Field(default=None, max_length=128)


class ResourceFeedbackDTO(BaseModel):
    id: UUID
    resource_id: UUID
    status: FeedbackStatus
    feedback_kinds: list[FeedbackKind]
    comment: str | None = None
    retry_workflow_run_id: UUID | None = None
    resulting_resource_id: UUID | None = None
    outcome: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class ResourceFeedbackSubmitResponse(BaseModel):
    feedback: ResourceFeedbackDTO
    workflow: dict[str, Any] | None = None


class ResourceLineageVersionDTO(BaseModel):
    resource_id: UUID
    version: int = Field(ge=1)
    parent_resource_id: UUID | None = None
    title: str
    status: str
    quality_score: float | None = None
    quality_delta: float | None = None
    changed_fields: list[str] = Field(default_factory=list)
    change_summary: str | None = None
    evidence_count: int = Field(ge=0)
    run_state: str
    source_kind: Literal["curated-demo", "external-preview", "real"]
    source_boundary: str
    created_at: datetime


class ResourceLineageDTO(BaseModel):
    lineage_root_id: UUID
    logical_key: str
    resource_type: str
    title: str
    knowledge_point: str | None = None
    current_resource_id: UUID
    versions: list[ResourceLineageVersionDTO] = Field(default_factory=list)


class StudentLearningLoopDTO(BaseModel):
    course_id: UUID
    candidates: list[PathReplanCandidateDTO] = Field(default_factory=list)
    path_versions: list[PathVersionDTO] = Field(default_factory=list)
    recommendations: list[ResourceRecommendationDTO] = Field(default_factory=list)
    feedback: list[ResourceFeedbackDTO] = Field(default_factory=list)
    resource_lineages: list[ResourceLineageDTO] = Field(default_factory=list)


__all__ = [
    "PathReplanCandidateDTO",
    "PathReplanCreateRequest",
    "PathReplanDecisionRequest",
    "PathVersionDTO",
    "ResourceFeedbackDTO",
    "ResourceFeedbackRequest",
    "ResourceFeedbackSubmitResponse",
    "ResourceLineageDTO",
    "ResourceRecommendationDecisionRequest",
    "StudentLearningLoopDTO",
]

File: frontend/src/app/features/course/studentLearningLoop.ts
import { apiGet, apiPost } from '@/lib/api';

export type ReplanDecision = 'accept' | 'defer' | 'revert';
export type RecommendationDecision = 'accept' | 'defer' | 'reject' | 'complete';
export type ResourceFeedbackKind =
  | 'too_difficult'
  | 'too_shallow'
  | 'missing_example'
  | 'want_diagram'
  | 'want_practice';

export type StudentLearningLoopCandidate = {
  id: string;
  status: 'pending' | 'deferred' | 'accepted' | 'reverted' | 'expired';
  source_version_no: number;
  accepted_version_no?: number | null;
  trigger_label: string;
  trigger_at?: string | null;
  reason_code: string;
  reason_text: string;
  affected_knowledge_point?: string | null;
  expected_minutes: number;
  changed_tasks: Array<{
    action: 'added' | 'retained';
    title: string;
    knowledge_point?: string | null;
    status: 'todo' | 'active' | 'done' | 'blocked';
    expected_minutes: number;
  }>;
  source_boundary: string;
  created_at: string;
  updated_at: string;
};

export type StudentLearningPathVersion = {
  id: string;
  version_no: number;
  kind: 'baseline' | 'replan' | 'revert';
  state: 'active' | 'historical';
  title: string;
  summary: string;
  diff: Record<string, unknown>;
  created_at: string;
};

export type StudentResourceRecommendation = {
  id: string;
  resource_id: string;
  title: string;
  resource_type: string;
  knowledge_point?: string | null;
  status: 'scheduled' | 'accepted' | 'deferred' | 'rejected' | 'superseded' | 'feedback_received' | 'completed';
  scheduled_at: string;
  rationale: string;
  source_boundary: string;
  created_at: string;
};

export type StudentResourceFeedback = {
  id: string;
  resource_id: string;
  status: 'submitted' | 'retry_requested' | 'regenerated' | 'provider_unavailable' | 'failed' | 'rejected';
  feedback_kinds: ResourceFeedbackKind[];
  comment?: string | null;
  retry_workflow_run_id?: string | null;
  resulting_resource_id?: string | null;
  outcome: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type StudentResourceLineage = {
  lineage_root_id: string;
  logical_key: string;
  resource_type: string;
  title: string;
  knowledge_point?: string | null;
  current_resource_id: string;
  versions: Array<{
    resource_id: string;
    version: number;
    parent_resource_id?: string | null;
    title: string;
    status: string;
    quality_score?: number | null;
    quality_delta?: number | null;
    changed_fields: string[];
    change_summary?: string | null;
    evidence_count: number;
    run_state: string;
    source_kind: 'curated-demo' | 'external-preview' | 'real';
    source_boundary: string;
    created_at: string;
  }>;
};

export type StudentLearningLoop = {
  course_id: string;
  candidates: StudentLearningLoopCandidate[];
  path_versions: StudentLearningPathVersion[];
  recommendations: StudentResourceRecommendation[];
  feedback: StudentResourceFeedback[];
  resource_lineages: StudentResourceLineage[];
};

export type ResourceFeedbackSubmitResponse = {
  feedback: StudentResourceFeedback;
  workflow?: Record<string, unknown> | null;
};

function base(courseId: string): string {
  return `/api/v1/courses/${encodeURIComponent(courseId)}/learning-loop`;
}

/** Current-user-only read model; it does not accept student, class, or evidence IDs. */
export function fetchStudentLearningLoop(courseId: string): Promise<StudentLearningLoop> {
  return apiGet<StudentLearningLoop>(base(courseId));
}

export function createStudentReplanCandidate(courseId: string): Promise<StudentLearningLoopCandidate> {
  return apiPost<StudentLearningLoopCandidate, Record<string, never>>(
    `${base(courseId)}/replan-candidates`,
    {},
  );
}

export function decideStudentReplanCandidate(
  courseId: string,
  candidateId: string,
  decision: ReplanDecision,
): Promise<StudentLearningLoopCandidate> {
  return apiPost<StudentLearningLoopCandidate, { decision: ReplanDecision }>(
    `${base(courseId)}/replan-candidates/${encodeURIComponent(candidateId)}/decision`,
    { decision },
  );
}

export function decideStudentRecommendation(
  courseId: string,
  recommendationId: string,
  decision: RecommendationDecision,
): Promise<StudentResourceRecommendation> {
  return apiPost<StudentResourceRecommendation, { decision: RecommendationDecision }>(
    `${base(courseId)}/recommendations/${encodeURIComponent(recommendationId)}/decision`,
    { decision },
  );
}

export function submitStudentResourceFeedback(
  courseId: string,
  resourceId: string,
  payload: {
    feedback_kinds: ResourceFeedbackKind[];
    comment?: string;
    recommendation_id?: string;
  },
): Promise<ResourceFeedbackSubmitResponse> {
  return apiPost<ResourceFeedbackSubmitResponse, typeof payload>(
    `${base(courseId)}/resources/${encodeURIComponent(resourceId)}/feedback`,
    payload,
  );
}
